File: rust/binette/src/compatibility.rs
use crate::error::SchemaError;
use crate::plan::{PlanError, reader_plan_for_bundle};
use crate::registry::SchemaRegistry;
use crate::schema::{SchemaBundle, TypeId, TypeRef};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CompatibilityStatus {
    Backward,
    Forward,
    Bidirectional,
    Incompatible,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum CompatibilityDirection {
    Backward,
    Forward,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CompatibilityFailure {
    pub direction: CompatibilityDirection,
    pub path: String,
    pub reason: CompatibilityFailureReason,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum CompatibilityFailureReason {
    UnknownWriterType { type_id: TypeId },
    UnknownReaderType { type_id: TypeId },
    UnboundTypeParameter { name: String },
    TypeMismatch { writer: TypeRef, reader: TypeRef },
    MissingReaderField { field: String },
    Unsupported { reason: &'static str },
    Schema { message: String },
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct CompatibilityReport {
    pub status: CompatibilityStatus,
    pub failures: Vec<CompatibilityFailure>,
}

// r[impl binette.compat.report]
pub fn compatibility_report(
    old: &SchemaBundle,
    new: &SchemaBundle,
) -> Result<CompatibilityReport, SchemaError> {
    let mut old_registry = SchemaRegistry::new();
    old_registry.install_bundle(old)?;
    let mut new_registry = SchemaRegistry::new();
    new_registry.install_bundle(new)?;

    let backward = reader_plan_for_bundle(&old.root, &old_registry, &new.root, &new_registry)
        .map(|_| ())
        .map_err(|error| compatibility_failure(CompatibilityDirection::Backward, error));
    let forward = reader_plan_for_bundle(&new.root, &new_registry, &old.root, &old_registry)
        .map(|_| ())
        .map_err(|error| compatibility_failure(CompatibilityDirection::Forward, error));

    let mut failures = Vec::new();
    let backward_ok = backward.is_ok();
    let forward_ok = forward.is_ok();

    if let Err(failure) = backward {
        failures.push(failure);
    }
    if let Err(failure) = forward {
        failures.push(failure);
    }

    let status = match (backward_ok, forward_ok) {
        (true, true) => CompatibilityStatus::Bidirectional,
        (true, false) => CompatibilityStatus::Backward,
        (false, true) => CompatibilityStatus::Forward,
        (false, false) => CompatibilityStatus::Incompatible,
    };

    Ok(CompatibilityReport { status, failures })
}

fn compatibility_failure(
    direction: CompatibilityDirection,
    error: PlanError,
) -> CompatibilityFailure {
    let (path, reason) = match error {
        PlanError::Schema(error) => (
            "$".to_owned(),
            CompatibilityFailureReason::Schema {
                message: error.to_string(),
            },
        ),
        PlanError::UnknownWriterType { path, type_id } => (
            path,
            CompatibilityFailureReason::UnknownWriterType { type_id },
        ),
        PlanError::UnknownReaderType { path, type_id } => (
            path,
            CompatibilityFailureReason::UnknownReaderType { type_id },
        ),
        PlanError::UnboundTypeParameter { path, name } => (
            path,
            CompatibilityFailureReason::UnboundTypeParameter { name },
        ),
        PlanError::TypeMismatch {
            path,
            writer,
            reader,
        } => (
            path,
            CompatibilityFailureReason::TypeMismatch { writer, reader },
        ),
        PlanError::MissingReaderField { path, field } => (
            path,
            CompatibilityFailureReason::MissingReaderField { field },
        ),
        PlanError::Unsupported { path, reason } => {
            (path, CompatibilityFailureReason::Unsupported { reason })
        }
    };

    CompatibilityFailure {
        direction,
        path,
        reason,
    }
}

File: typescript/binette/src/compatibility.ts
import {
  PlanError,
  readerPlanForBundle,
  type PlanErrorKind,
} from "./plan.js";
import { SchemaRegistry } from "./registry.js";
import { type SchemaBundle } from "./schema.js";

export type CompatibilityStatus =
  | "backward"
  | "forward"
  | "bidirectional"
  | "incompatible";

export type CompatibilityDirection = "backward" | "forward";

export type CompatibilityFailure = {
  direction: CompatibilityDirection;
  path: string;
  reason: CompatibilityFailureReason;
};

export type CompatibilityFailureReason =
  | { kind: PlanErrorKind; detail: unknown }
  | { kind: "schema"; message: string };

export type CompatibilityReport = {
  status: CompatibilityStatus;
  failures: CompatibilityFailure[];
};

// r[impl binette.compat.report]
export function compatibilityReport(
  oldBundle: SchemaBundle,
  newBundle: SchemaBundle,
): CompatibilityReport {
  const oldRegistry = new SchemaRegistry();
  oldRegistry.installBundle(oldBundle);
  const newRegistry = new SchemaRegistry();
  newRegistry.installBundle(newBundle);

  const backward = directionResult(
    "backward",
    () =>
      readerPlanForBundle(
        oldBundle.root,
        oldRegistry,
        newBundle.root,
        newRegistry,
      ),
  );
  const forward = directionResult(
    "forward",
    () =>
      readerPlanForBundle(
        newBundle.root,
        newRegistry,
        oldBundle.root,
        oldRegistry,
      ),
  );
  const failures = [backward.failure, forward.failure].filter(
    (failure): failure is CompatibilityFailure => failure !== null,
  );

  let status: CompatibilityStatus;
  if (backward.ok && forward.ok) {
    status = "bidirectional";
  } else if (backward.ok) {
    status = "backward";
  } else if (forward.ok) {
    status = "forward";
  } else {
    status = "incompatible";
  }

  return { status, failures };
}

function directionResult(
  direction: CompatibilityDirection,
  plan: () => unknown,
): { ok: boolean; failure: CompatibilityFailure | null } {
  try {
    plan();
    return { ok: true, failure: null };
  } catch (error) {
    return { ok: false, failure: compatibilityFailure(direction, error) };
  }
}

function compatibilityFailure(
  direction: CompatibilityDirection,
  error: unknown,
): CompatibilityFailure {
  if (error instanceof PlanError) {
    return {
      direction,
      path: error.path,
      reason: { kind: error.kind, detail: error.detail },
    };
  }
  return {
    direction,
    path: "$",
    reason: { kind: "schema", message: String(error) },
  };
}
